File: backend/app/schemas/hierarchy.py
from typing import List, Optional

from pydantic import BaseModel, Field


class HierarchyNodeInput(BaseModel):
    name: str
    node_type: str = Field(..., description="goal|criteria|alternative")
    children: Optional[List["HierarchyNodeInput"]] = None


class HierarchyNodeOutput(BaseModel):
    id: str
    name: str
    node_type: str
    parent_id: Optional[str]
    sort_order: int


class HierarchyNodeCreateRequest(BaseModel):
    name: str
    node_type: str
    parent_id: Optional[str] = None
    sort_order: Optional[int] = None


class HierarchyNodeUpdateRequest(BaseModel):
    name: Optional[str] = None
    node_type: Optional[str] = None
    parent_id: Optional[str] = None
    sort_order: Optional[int] = None


class HierarchyTreeNode(BaseModel):
    id: str
    name: str
    node_type: str
    parent_id: Optional[str]
    sort_order: int
    children: Optional[List["HierarchyTreeNode"]] = None


HierarchyNodeInput.model_rebuild()
HierarchyTreeNode.model_rebuild()

File: frontend/src/types/hierarchy.ts
/**
 * 계층 구조 공통 타입·상수 (구조도·편집 행 공통화용)
 */

export type NodeType = "goal" | "criteria" | "alternative";

/** 구조도·읽기 전용에 사용. id 없이 name, node_type, children 만 있으면 됨 */
export interface HierarchyNodeBase {
  name: string;
  node_type: string;
  children?: HierarchyNodeBase[];
}

export const nodeTypeOptions: { value: NodeType; label: string }[] = [
  { value: "goal", label: "목표" },
  { value: "criteria", label: "기준" },
  { value: "alternative", label: "대안" },
];

export const typeLabel: Record<NodeType, string> = {
  goal: "목표",
  criteria: "기준",
  alternative: "대안",
};
